src/prof/src/lib.rs

```diff
@@ -35,6 +35,181 @@ pub struct StackProfile {
     stacks: Vec<(WeightedStack, Option<usize>)>,
 }
 
+/// Gets the GNU build IDs for all loaded images, including the main
+/// program binary as well as all dynamically loaded libraries.
+/// Intended to be useful for profilers, who can use the supplied IDs
+/// to symbolicate stack traces offline.
+///
+/// Uses `dl_iterate_phdr` to walk the program headers of all images,
+/// and iterates over them looking for note
+/// segments. Then searches the discovered note segments for a note of type
+/// `NT_GNU_BUILD_ID` (aka "3") and name "GNU\0".
+///
+/// SAFETY: This function is written in a hilariously unsafe way: it involves
+/// following pointers to random parts of memory, and then assuming
+/// that particular structures can be found there.
+/// However, it was written by carefully reading `man dl_iterate_phdr`
+/// and `man elf`, and is thus intended to be relatively safe for callers to use.
+/// Assuming I haven't written any bugs (and that the documentation is correct),
+/// the only known safety requirements are:
+///
+/// (1) It must not be called multiple times concurrently, as `dl_iterate_phdr`
+/// is not documented as being thread-safe.
+/// (2) The running binary must be in ELF format and running on Linux.
+#[cfg(not(target_os = "macos"))]
+pub unsafe fn all_build_ids(
+) -> Result<std::collections::HashMap<std::path::PathBuf, Vec<u8>>, anyhow::Error> {
+    // local imports to avoid polluting the namespace for macOS builds
+    use std::collections::hash_map::Entry;
+    use std::collections::HashMap;
+    use std::ffi::{c_int, CStr, OsStr};
+    use std::os::unix::ffi::OsStrExt;
+    use std::path::{Path, PathBuf};
+
+    use mz_ore::cast::CastFrom;
+
+    use anyhow::Context;
+    use libc::{dl_iterate_phdr, dl_phdr_info, size_t, Elf64_Word, PT_NOTE};
+
+    struct CallbackState {
+        map: HashMap<PathBuf, Vec<u8>>,
+        is_first: bool,
+        fatal_error: Option<anyhow::Error>,
+    }
+    extern "C" fn iterate_cb(info: *mut dl_phdr_info, _size: size_t, data: *mut c_void) -> c_int {
```

Comment:
I guess to be totally correct you'd bail if `_size` is less than the end of the last member of `info` you use, but this code only uses the base members that should always be there so I think this is fine.

Reply:
(It's not about broken structs, it's about compatibility.  The reason for passing the size argument is so that programs can ensure that the members they use are present, and future changes to the struct are guaranteed to only extend it.  Older versions of glibc will pass a smaller value of size than newer ones, for example.)

Reply:
I see, but shouldn't this be impossible?

I.e., I'd expect that the `libc` binding crate is set up in such a way that it wouldn't let you link against a version of glibc that would cause it to return partially uninitialized structs to Rust code. (I haven't actually verified whether this is the case)

Reply:
At least in our current builds, libc is dynamically linked and can be any version at runtime.  The same principle applies for a lot of syscall interfaces, where again static verification is impossible.  I mention this only because this mechanism is interesting (size as version information) and failing to check (or maybe worse, asserting equality) is a common source of bugs.

Reply:
There are symbol versioning mechanisms for requiring a minimum glibc version, at least, but I don't know if the libc crate uses them.

Reply:
I'd consider that a bug in the libc crate! But anyway, I don't think it can happen to us, because we control the environment we run on, and as you said, the fields I'm reading will always be there.